dcore/configuration.py

```diff
@@ -0,0 +1,81 @@
+"""
+Some simple tools for making model configuration nicer.
+"""
+
+
+class Configuration(object):
+    def __init__(self, **kwargs):
+
+        self.__setattr__ = self._proto__setattr__
+        for name, value in kwargs.iteritems():
+            print name, value
+            self.__setattr__(name, value)
+
+    def _proto__setattr__(self, name, value):
+        """Cause setting an unknown attribute to be an error"""
+        self.__getattribute__(name)
+        object.__setattr__(self, name, value)
+
+
+class TimesteppingParameters(Configuration):
+
+    """
+    Timestepping parameters for dcore
+    """
+
+    def __init__(self, **kwargs):
+        self.dt = None
+        self.alpha = 0.5
+        self.maxk = 2
+        self.maxi = 2
+
+        super(TimesteppingParameters, self).__init__(**kwargs)
+
+
+class OutputParameters(Configuration):
+
+    """
+    Output parameters for dcore
+    """
+
+    def __init__(self, **kwargs):
+        self.Verbose = False
+        self.dumpfreq = 10
+        self.dumplist = (True,True,True)
+
+        super(OutputParameters, self).__init__(**kwargs)
+
+
+class CompressibleParameters(Configuration):
+
+    """
+    Physical parameters for 3d Compressible Euler
+    """
+
+    def __init__(self, **kwargs):
+        self.g = 9.81
+        self.N = 0.01
+        self.cp = 1004.5
+        self.R_d = 287.
+        self.p_0 = 1000.0*1000.0
+        self.kappa = 2.0/7.0
+        self.k = None
+        self.Omega = None
+
+        super(CompressibleParameters, self).__init__(**kwargs)
```

Comment:
If I understand this class correctly, I would use:

```python
class Configuration(object):

    def __init__(self, **kwargs):
        for name, value in kwargs.iteritems():
            self.__setattr__(name, value)

    def __setattr__(self, name, value):
        if not hasattr(self, name):
            raise AttributeError("'%s' object has no attribute '%s'" % (type(self).__name__, name))
        object.__setattr__(self, name, value)
```

Now the inherited configuration object don't have an init, they just provide a default value for each of the parameters you're allowed to set:

```python
class SubConfig(Configuration):
    a = None
    g = 9.81
```

Now, when I make a `SubConfig`, I can either just select the defaults:

```python
conf = SubConfig()
```

Or else override some defaults:
```python
conf = SubConfig(a=55)
```

If I try and set an "unsupported" parameter, then I get an attribute error:

```python
x.c = 14321

Truncated Traceback (Use C-c C-x to view full TB):
<ipython-input-31-d5c9fd37ecc0> in __setattr__(self, name, value)
      2     def __setattr__(self, name, value):
      3         if not hasattr(self, name):
----> 4             raise AttributeError("'%s' object has no attribute '%s'" % (type(self).__name__, name))
      5         object.__setattr__(self, name, value)
      6 

AttributeError: 'SubConfig' object has no attribute 'c'
```

I think this makes the code somewhat simpler, because you don't need init and so on.

Certainly, that's what the example below suggests the mechanism is.